src/streaming/zarr.dimension.cpp

```diff
@@ -6,8 +6,34 @@ ArrayDimensions::ArrayDimensions(std::vector<ZarrDimension>&& dims,
                                  ZarrDataType dtype)
   : dims_(std::move(dims))
   , dtype_(dtype)
+  , chunks_per_shard_(1)
+  , number_of_shards_(1)
+  , bytes_per_chunk_(zarr::bytes_of_type(dtype))
+  , number_of_chunks_in_memory_(1)
 {
     EXPECT(dims_.size() > 2, "Array must have at least three dimensions.");
+
+    for (const auto& d : dims_) {
+        bytes_per_chunk_ *= d.chunk_size_px;
+    }
+
+    for (auto i = 1; i < ndims(); ++i) {
+        number_of_chunks_in_memory_ *= zarr::chunks_along_dimension(dims_[i]);
+    }
+
+    for (const auto& dim : dims_) {
+        chunks_per_shard_ *= dim.shard_size_chunks;
+    }
+
+    for (auto i = 1; i < ndims(); ++i) {
+        const auto& dim = dims_[i];
+        number_of_shards_ *= zarr::shards_along_dimension(dim);
+    }
```

Comment:
I think that you can collapse all of these loops into one unless I'm missing something?

```cpp
for (size_t i = 0; i < dims_.size(); ++i) {
    const auto& dim = dims_[i];
    bytes_per_chunk_ *= dim.chunk_size_px;
    chunks_per_shard_ *= dim.shard_size_chunks;
    if (i > 0) {
        number_of_chunks_in_memory_ *= zarr::chunks_along_dimension(dim);
        number_of_shards_ *= zarr::shards_along_dimension(dim);
    }
}
```